exec/java-exec/src/main/java/org/apache/drill/exec/record/BatchSchema.java

```diff
@@ -113,10 +115,45 @@ public boolean equals(Object obj) {
     } else if (!fields.equals(other.fields)) {
       return false;
     }
+    for (int i = 0; i < fields.size(); i++) {
+      MajorType t1 = fields.get(i).getType();
+      MajorType t2 = other.fields.get(i).getType();
+      if (t1 == null) {
+        if (t2 != null) {
+          return false;
+        }
+      } else {
+        if (!majorTypeEqual(t1, t2)) {
+          return false;
+        }
+      }
+    }
     if (selectionVectorMode != other.selectionVectorMode) {
       return false;
     }
     return true;
   }
 
+  /**
+   * We treat fields with same set of Subtypes as equal, even if they are in a different order
+   * @param t1
+   * @param t2
+   * @return
+   */
+  private boolean majorTypeEqual(MajorType t1, MajorType t2) {
+    if (t1.equals(t2)) {
+      return true;
+    }
+    if (!t1.getMinorType().equals(t2.getMinorType())) {
+      return false;
+    }
+    if (!t1.getMode().equals(t2.getMode())) {
+      return false;
+    }
+    if (!Sets.newHashSet(t1.getSubTypeList()).equals(Sets.newHashSet(t2.getSubTypeList()))) {
```

Comment:
Should we recursively check that the subtype of subtypes are also set equal as opposed to your current direct equality (below the second level)